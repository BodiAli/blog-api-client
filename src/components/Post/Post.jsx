import { useEffect, useReducer, useState } from "react";
import { useNavigate, useParams, useOutletContext, Link } from "react-router";
import { toast } from "react-toastify";
import { format, formatDistanceToNow } from "date-fns";
import parse from "html-react-parser";
import CommentCard from "../CommentCard/CommentCard";
import Loader from "../Loader/Loader";
import noImage from "../../assets/images/no-image.svg";
import anonymousImage from "../../assets/images/anonymous.jpg";
import styles from "./Post.module.css";
import FormButton from "../FormButton/FormButton";

export default function Post() {
  const [post, dispatch] = useReducer(reducer, null);
  const [loading, setLoading] = useState(true);
  const [invalidInputErrors, setInvalidInputErrors] = useState([]);
  const navigate = useNavigate();
  const { postId } = useParams();
  const { token, user } = useOutletContext();

  async function handleLikePost() {
    if (!post) return;

    try {
      const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts/${post.id}/like`, {
        method: "PATCH",
        headers: {
          Authorization: token,
        },
      });

      if (!res.ok) {
        if (post.postLiked) {
          throw new Error("Failed to remove like please try again later");
        } else {
          throw new Error("Failed to like post please try again later");
        }
      }

      if (post.postLiked) {
        toast.success("Like removed successfully!");
      } else {
        toast.success("Post liked successfully!");
      }

      dispatch({
        type: "update-post-like",
      });
    } catch (error) {
      toast.error(error.message);
    }
  }

  async function handleCreateComment(e) {
    e.preventDefault();

    if (!user) {
      toast.info("You need to login to create a comment");
      return;
    }

    const formData = new FormData(e.currentTarget);

    const content = formData.get("content");

    try {
      const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts/${post.id}/comments`, {
        method: "POST",
        headers: {
          Authorization: token,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ content }),
      });

      if (!res.ok) {
        if (res.status === 400) {
          const { errors } = await res.json();

          setInvalidInputErrors(errors);
          return;
        }

        throw new Error("Failed to create comment please try again later");
      }

      const { msg, comment } = await res.json();

      dispatch({
        type: "create-comment",
        comment,
      });
      toast.success(msg);
    } catch (error) {
      toast.error(error.message);
    }
  }

  useEffect(() => {
    async function fetchPost() {
      try {
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts/${postId}`, {
          headers: {
            Authorization: token,
          },
        });

        if (!res.ok) {
          if (res.status === 404) {
            const { error } = await res.json();
            throw new Error(error);
          }
          throw new Error("Failed to fetch post please try again later");
        }

        const fetchedPost = await res.json();

        dispatch({ type: "set-post", post: fetchedPost });
      } catch (error) {
        toast.error(error.message);
        navigate(-1, { viewTransition: true });
      } finally {
        setLoading(false);
      }
    }
    fetchPost();
  }, [postId, navigate, token]);

  if (loading) return <Loader />;

  return (
    <main className={styles.main}>
      <div className={styles.textContainer1}>
        <h2 className={styles.title}>{post?.title}</h2>
        <p className={styles.date}>
          <strong>Created at: </strong>
          {post && format(post.createdAt, "y/M/d, H:m:s")}
        </p>
        <p className={styles.date}>
          <strong>Last updated: </strong>
          {post && formatDistanceToNow(post.updatedAt, { addSuffix: true, includeSeconds: true })}
        </p>
      </div>
      <div className={styles.textContainer2}>
        <p>Topics:</p>
        <div>
          {post?.Topics.length === 0 ? (
            <span>This post has no topics</span>
          ) : (
            post?.Topics.map((topic, i) => {
              return <p key={i}>{topic.name}</p>;
            })
          )}
        </div>
      </div>
      <div className={styles.postContent}>
        <img src={post?.imgUrl ? post?.imgUrl : noImage} alt="Post cover" />
        <div className={styles.postContentHtml}>{post && parse(post.content)}</div>
      </div>
      <hr />
      <div className={styles.likeContainer}>
        <p className={styles.likes}>
          <strong>Likes:</strong> {post?.likes}
        </p>
        {!user ? (
          <p className={styles.loginMessage} data-testid="login-notice">
            Please{" "}
            <Link to="log-in" viewTransition>
              login
            </Link>{" "}
            to be able to like posts
          </p>
        ) : (
          <button
            data-testid="like-button"
            onClick={handleLikePost}
            className={`${styles.likeButton} ${post?.postLiked ? styles.liked : ""}`}
          ></button>
        )}
      </div>
      <div className={styles.commentsContainer}>
        <h3>Comments</h3>
        <div className={styles.formContainer}>
          <ul className={styles.errors}>
            {invalidInputErrors.map((error, i) => {
              return <li key={i}>{error.msg}</li>;
            })}
          </ul>
          <form onSubmit={handleCreateComment} className={styles.createComment}>
            <img
              src={user.Profile.profileImgUrl ? user.Profile.profileImgUrl : anonymousImage}
              alt={`${user.firstName} ${user.lastName}'s profile picture`}
            />
            <input type="text" name="content" placeholder="Comment on post" required />
            <FormButton>Create</FormButton>
          </form>
        </div>
        <div className={styles.commentsCardContainer}>
          {post?.Comments.map((comment) => {
            return (
              <CommentCard key={comment.id} comment={comment} user={user} dispatch={dispatch} token={token} />
            );
          })}
        </div>
      </div>
    </main>
  );
}

function reducer(post, action) {
  switch (action.type) {
    case "set-post": {
      return action.post;
    }
    case "update-post-like": {
      return { ...post, postLiked: !post.postLiked, likes: post.postLiked ? post.likes - 1 : post.likes + 1 };
    }
    case "create-comment": {
      const updatedComments = [action.comment, ...post.Comments];

      return { ...post, Comments: updatedComments };
    }
    case "update-comment-like": {
      const updatedComments = post.Comments.map((comment) => {
        if (comment.id === action.comment.id) {
          return {
            ...comment,
            commentLiked: !comment.commentLiked,
            likes: comment.commentLiked ? comment.likes - 1 : comment.likes + 1,
          };
        }
        return comment;
      });
      return { ...post, Comments: updatedComments };
    }
    case "update-comment-content": {
      const updatedComments = post.Comments.map((comment) => {
        if (comment.id === action.comment.id) {
          return { ...comment, content: action.content };
        }
        return comment;
      });

      return { ...post, Comments: updatedComments };
    }
    case "delete-comment": {
      const updatedComments = post.Comments.filter((comment) => {
        return comment.id !== action.comment.id;
      });
      return { ...post, Comments: updatedComments };
    }
    default: {
      throw new Error(`Unknown action ${action.type}`);
    }
  }
}
