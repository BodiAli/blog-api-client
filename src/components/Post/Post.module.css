.main {
  padding: 1rem;
  font-family: "roboto", Arial, sans-serif;
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.textContainer1 {
  display: flex;
  align-items: center;
}

.textContainer1 > * {
  flex: 1;
}

.textContainer1 > h2 {
  font-size: 1.5rem;
}

.textContainer1 strong {
  font-weight: 500;
}

.textContainer1 > .date {
  text-align: center;
  color: #0000009c;
}

.date:last-child {
  text-align: end;
}

.textContainer2 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.textContainer2 > p {
  font-weight: 500;
}

.textContainer2 > div {
  display: flex;
  align-items: start;
  flex-wrap: wrap;
  gap: 1rem;
}

.textContainer2 > div > p {
  font-weight: 500;
  color: rgba(115, 113, 113, 0.761);
  background-color: rgba(128, 128, 128, 0.246);
  border-radius: 5px;
  padding: 0.3rem 0.6rem;
}

.postContent {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.postContent > img {
  width: 80%;
  border-radius: 5px;
  align-self: center;
}

.postContentHtml > * {
  all: revert;
}

.likeContainer {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.likeContainer strong {
  font-weight: 500;
}

.likeButton {
  background-color: black;
  mask-image: url("../../assets/images/heart.svg");
  mask-repeat: no-repeat;
  mask-size: contain;
  width: 25px;
  aspect-ratio: 1 / 1;
  cursor: pointer;
}

.likeButton.liked {
  background-color: crimson;
}

.loginMessage a {
  color: #000;
}

.main > hr {
  width: 100%;
}

.commentsContainer {
  display: grid;
  grid-template-columns: 1fr;
  padding: 0.5rem;
  border-radius: 5px;
  gap: 2rem;
}

.commentsContainer > h3 {
  font-size: 1.4rem;
  justify-self: center;
}

.formContainer {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

.errors {
  color: crimson;
  list-style: disc;
  list-style-position: inside;
  justify-self: center;
  line-height: 1.3;
}

.createComment {
  display: grid;
  grid-template-columns: auto 1fr auto;
  align-items: center;
  gap: 1rem;
}

.createComment > img {
  width: 80px;
  aspect-ratio: 1 / 1;
  object-fit: cover;
  border-radius: 50%;
}

.createComment > input {
  border: none;
  border-radius: 5px;
  box-shadow: 0 0px 5px gray;
  padding: 1rem 0.5rem;
}

.commentsContainer > .commentsCardContainer {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

@media (max-width: 770px) {
  .textContainer1 {
    flex-direction: column;
    gap: 1.5rem;
    align-items: start;
    margin-top: 1rem;
  }

  .textContainer1 > h2 {
    align-self: center;
  }

  .postContent > img {
    width: 100%;
  }
}

@media (max-width: 500px) {
  .createComment {
    grid-template-columns: 1fr;
  }
  .createComment > img {
    justify-self: center;
  }
}
