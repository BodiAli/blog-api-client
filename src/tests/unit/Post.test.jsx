import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createRoutesStub, useOutletContext } from "react-router";
import { toast } from "react-toastify";
import { beforeEach, expect, vi } from "vitest";
import Post from "../../components/Post/Post";

vi.stubGlobal("fetch", vi.fn());

vi.mock("date-fns", async (importOriginal) => {
  const result = await importOriginal();

  return {
    ...result,
    formatDistanceToNow: () => "1 day ago",
  };
});

vi.mock("react-toastify", () => {
  return {
    toast: {
      error: vi.fn(),
      success: vi.fn(),
      info: vi.fn(),
    },
  };
});

const fakeUser = {
  id: 1,
  firstName: "Bodi",
  lastName: "Ali",
  Profile: {
    profileImgUrl: "imageUrl",
  },
};

const navigateMock = vi.fn();

vi.mock("react-router", async (importOriginal) => {
  const result = await importOriginal();

  return {
    ...result,
    useOutletContext: vi.fn(),
    useNavigate: () => navigateMock,
  };
});

const fakePost = {
  id: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
  userId: 1,
  title: "Post title",
  content: "Post content",
  published: true,
  likes: 2,
  createdAt: "2025-05-01T10:44:36.648Z",
  updatedAt: "2025-05-03T04:45:02.965Z",
  imgUrl: "imageUrl",
  User: {
    id: 1,
    firstName: "Bodi",
    lastName: "Ali",
    Profile: {
      profileImgUrl: "imageUrl",
    },
  },
  Topics: [{ name: "Topic 1" }, { name: "Topic 2" }],
  postLiked: false,
  Comments: [
    {
      id: "7397319e-58ef-4e42-bc94-aa8788db4090",
      userId: 1,
      postId: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
      content: "Comment content",
      likes: 1,
      createdAt: "2025-05-03T07:51:49.273Z",
      updatedAt: "2025-05-04T06:19:03.092Z",
      User: {
        id: 1,
        firstName: "Bodi",
        lastName: "Ali",
        Profile: {
          profileImgUrl: "imageUrl",
        },
      },
      commentLiked: false,
    },
    {
      id: "5432119e-58ef-4e42-bc94-aa8788db2021",
      userId: 2,
      postId: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
      content: "Comment content 2",
      likes: 1,
      createdAt: "2025-05-03T07:51:49.273Z",
      updatedAt: "2025-05-04T06:19:03.092Z",
      User: {
        id: 1,
        firstName: "John",
        lastName: "Doe",
        Profile: {
          profileImgUrl: "imageUrl",
        },
      },
      commentLiked: false,
    },
  ],
};

beforeEach(() => {
  fetch.mockImplementation(() => {
    return Promise.resolve({
      ok: true,
      status: 200,
      json: () => Promise.resolve(fakePost),
    });
  });

  useOutletContext.mockImplementation(() => ({ token: "token", user: fakeUser }));

  vi.clearAllMocks();
});

describe("Post component", () => {
  test("Should render post and comment data with expected buttons and inputs", async () => {
    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    let container;

    await waitFor(() => {
      const result = render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);

      container = result.container;
    });

    expect(container).toMatchSnapshot();
  });

  test("When component is mounted it should call fetch with expected postId and Authorization header as the token", async () => {
    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    const fetchCall = fetch.mock.calls[0];

    expect(fetchCall[0]).toMatch("57c20dfb-8135-42eb-8abe-0bffbaffd668");
    expect(fetchCall[1]).toStrictEqual({ headers: { Authorization: "token" } });
  });

  test("if user is not authenticated, component should render a login notice for liking post", async () => {
    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    useOutletContext.mockImplementation(() => ({ token: "token", user: null }));

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    expect(screen.getByTestId("login-notice")).toBeInTheDocument();
    expect(screen.queryByTestId("like-button")).not.toBeInTheDocument();
  });

  test("should increment post likes to 3 when clicking like button and assign liked class to like button", async () => {
    const user = userEvent.setup();

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    await user.click(screen.getByTestId("like-button"));

    expect(screen.getByTestId("likes-count")).toHaveTextContent("Likes: 3");
    expect(screen.getByTestId("like-button")).toHaveClass(/liked/i);
    expect(toast.success).toHaveBeenCalledWith("Post liked successfully!");
  });

  test("should decrement post likes to 1 when clicking unlike button and remove liked class from like button", async () => {
    const user = userEvent.setup();

    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ ...fakePost, postLiked: true }),
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    await user.click(screen.getByTestId("like-button"));

    expect(screen.getByTestId("likes-count")).toHaveTextContent("Likes: 1");
    expect(screen.getByTestId("like-button")).not.toHaveClass(/liked/i);
    expect(toast.success).toHaveBeenCalledWith("Like removed successfully!");
  });

  test("if fetch response is not ok because post is not found, component should call navigate and error toast with expected arguments", async () => {
    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: false,
        status: 404,
        json: () =>
          Promise.resolve({
            error: "Post not found! it may have been moved, deleted or it might have never existed.",
          }),
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/invalidParam"]} />);
    });

    expect(navigateMock).toHaveBeenCalledWith(-1, { viewTransition: true });
    expect(toast.error).toHaveBeenCalledWith(
      "Post not found! it may have been moved, deleted or it might have never existed."
    );
  });

  test("When create comment form is submitted and user is not authenticated it should info a toast", async () => {
    const user = userEvent.setup();

    useOutletContext.mockImplementation(() => ({ token: "", user: null }));

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    const commentInput = screen.getByPlaceholderText("Comment on post");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(commentInput, "Created comment");
    await user.click(createButton);

    expect(toast.info).toBeCalledWith("You need to login to create a comment");
  });

  test("When create comment form is submitted it should create a new comment", async () => {
    const user = userEvent.setup();

    // Implement two responses one for useEffect and one for handleCreateComment
    fetch
      .mockImplementationOnce(() => {
        return Promise.resolve({
          ok: true,
          status: 200,
          json: () => Promise.resolve(fakePost),
        });
      })
      .mockImplementationOnce(() => {
        return Promise.resolve({
          ok: true,
          status: 201,
          json: () =>
            Promise.resolve({
              comment: {
                id: "1234-1234-1234-1234-1234",
                userId: 1,
                postId: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
                content: "Created comment",
                likes: 1,
                createdAt: "2025-05-03T07:51:49.273Z",
                updatedAt: "2025-05-04T06:19:03.092Z",
                User: {
                  id: 1,
                  firstName: "Bodi",
                  lastName: "Ali",
                  Profile: {
                    profileImgUrl: "imageUrl",
                  },
                },
                commentLiked: false,
              },
              msg: "Comment created successfully!",
            }),
        });
      });

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    const commentInput = screen.getByPlaceholderText("Comment on post");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(commentInput, "Created comment");
    await user.click(createButton);

    expect(screen.getByText("Created comment")).toBeInTheDocument();
    expect(toast.success).toHaveBeenCalledWith("Comment created successfully!");
  });

  test("If form is submitted with invalid input component should render invalid input error returned from server", async () => {
    const user = userEvent.setup();

    // Implement two responses one for useEffect and one for handleCreateComment
    fetch
      .mockImplementationOnce(() => {
        return Promise.resolve({
          ok: true,
          status: 200,
          json: () => Promise.resolve(fakePost),
        });
      })
      .mockImplementationOnce(() => {
        return Promise.resolve({
          ok: false,
          status: 400,
          json: () => Promise.resolve({ errors: [{ msg: "Comment can not be empty." }] }),
        });
      });

    const Stub = createRoutesStub([
      {
        path: "/posts/:postId",
        Component: () => <Post />,
      },
    ]);

    await waitFor(() => {
      render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
    });

    const commentInput = screen.getByPlaceholderText("Comment on post");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(commentInput, "Created comment");
    await user.click(createButton);

    expect(screen.getByRole("listitem")).toHaveTextContent("Comment can not be empty.");
  });

  // test("KOKO", async () => {
  //   const user = userEvent.setup();

  //   fetch
  //     .mockImplementationOnce(() => {
  //       return Promise.resolve({
  //         ok: true,
  //         status: 200,
  //         json: () => Promise.resolve(fakePost),
  //       });
  //     })
  //     .mockImplementationOnce(() => {
  //       return Promise.resolve({
  //         ok: false,
  //         status: 400,
  //         json: () => Promise.resolve({ errors: [{ msg: "Comment can not be empty." }] }),
  //       });
  //     });

  //   const Stub = createRoutesStub([
  //     {
  //       path: "/posts/:postId",
  //       Component: () => <Post />,
  //     },
  //   ]);

  //   await waitFor(() => {
  //     render(<Stub initialEntries={["/posts/57c20dfb-8135-42eb-8abe-0bffbaffd668"]} />);
  //   });
  // });
});
