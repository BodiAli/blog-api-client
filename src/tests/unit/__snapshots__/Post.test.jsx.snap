// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Post component > Should render post and comment data with expected buttons and inputs 1`] = `
<div>
  <main
    class="_main_53c57f"
  >
    <div
      class="_textContainer1_53c57f"
    >
      <h2
        class="_title_53c57f"
      >
        Post title
      </h2>
      <p
        class="_date_53c57f"
      >
        <strong>
          Created at: 
        </strong>
        May 1, 2025, 13:44:36
      </p>
      <p
        class="_date_53c57f"
      >
        <strong>
          Last updated: 
        </strong>
        1 day ago
      </p>
    </div>
    <div
      class="_textContainer2_53c57f"
    >
      <p>
        Topics:
      </p>
      <div>
        <p>
          Topic 1
        </p>
        <p>
          Topic 2
        </p>
      </div>
    </div>
    <div
      class="_postContent_53c57f"
    >
      <img
        alt="Post cover"
        src="imageUrl"
      />
      <div
        class="_postContentHtml_53c57f"
      >
        Post content
      </div>
    </div>
    <hr />
    <div
      class="_likeContainer_53c57f"
    >
      <p
        class="_likes_53c57f"
        data-testid="likes-count"
      >
        <strong>
          Likes:
        </strong>
         
        2
      </p>
      <button
        class="_likeButton_53c57f "
        data-testid="like-button"
      />
    </div>
    <div
      class="_commentsContainer_53c57f"
    >
      <h3>
        Comments
      </h3>
      <div
        class="_formContainer_53c57f"
      >
        <ul
          class="_errors_53c57f"
        />
        <form
          class="_createComment_53c57f"
        >
          <img
            alt="Bodi Ali's profile picture"
            src="imageUrl"
          />
          <input
            name="content"
            placeholder="Comment on post"
            required=""
            type="text"
          />
          <button
            class="_button_64628a "
            type="submit"
          >
            Create
          </button>
        </form>
      </div>
      <div
        class="_commentsCardContainer_53c57f"
      >
        <div
          class="_card_ec5f2a"
        >
          <div
            class="_container1_ec5f2a"
          >
            <div
              class="_commentUserInfo_ec5f2a"
            >
              <img
                alt="Bodi Ali's profile picture"
                src="imageUrl"
              />
              <p
                class="_name_ec5f2a"
              >
                Bodi
                 
                Ali
              </p>
              <p
                class="_createdAt_ec5f2a"
              >
                1 day ago
              </p>
            </div>
            <div
              class="_dropdown_ec5f2a"
            >
              <button
                class="_dropBtn_ec5f2a"
              >
                <img
                  alt="options"
                  src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M12,16A2,2%200%200,1%2014,18A2,2%200%200,1%2012,20A2,2%200%200,1%2010,18A2,2%200%200,1%2012,16M12,10A2,2%200%200,1%2014,12A2,2%200%200,1%2012,14A2,2%200%200,1%2010,12A2,2%200%200,1%2012,10M12,4A2,2%200%200,1%2014,6A2,2%200%200,1%2012,8A2,2%200%200,1%2010,6A2,2%200%200,1%2012,4Z'%20/%3e%3c/svg%3e"
                />
              </button>
              <div
                class="_dropdownContent_ec5f2a"
              >
                <div
                  class="_folderEdit_ec5f2a"
                >
                  <img
                    alt="Edit comment"
                    src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3%2017.15,3.1%2016.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65%2021.1,6%2020.71,5.63L18.37,3.29C18.17,3.09%2017.92,3%2017.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z'%20/%3e%3c/svg%3e"
                  />
                  <button>
                    Edit
                  </button>
                </div>
                <div
                  class="_folderDelete_ec5f2a"
                >
                  <img
                    alt="delete comment"
                    src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M9,3V4H4V6H5V19A2,2%200%200,0%207,21H17A2,2%200%200,0%2019,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z'%20/%3e%3c/svg%3e"
                  />
                  <button>
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="_container2_ec5f2a"
          >
            <p>
              Comment content
            </p>
          </div>
          <div
            class="_likeContainer_ec5f2a"
          >
            <p
              class="_likes_ec5f2a"
            >
              <strong>
                Likes:
              </strong>
               
              1
            </p>
            <button
              class="_likeButton_ec5f2a "
              data-testid="like-button-comment"
            />
          </div>
        </div>
        <div
          class="_card_ec5f2a"
        >
          <div
            class="_container1_ec5f2a"
          >
            <div
              class="_commentUserInfo_ec5f2a"
            >
              <img
                alt="John Doe's profile picture"
                src="imageUrl"
              />
              <p
                class="_name_ec5f2a"
              >
                John
                 
                Doe
              </p>
              <p
                class="_createdAt_ec5f2a"
              >
                1 day ago
              </p>
            </div>
          </div>
          <div
            class="_container2_ec5f2a"
          >
            <p>
              Comment content 2
            </p>
          </div>
          <div
            class="_likeContainer_ec5f2a"
          >
            <p
              class="_likes_ec5f2a"
            >
              <strong>
                Likes:
              </strong>
               
              1
            </p>
            <button
              class="_likeButton_ec5f2a "
              data-testid="like-button-comment"
            />
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
